File: packages/sdk-python/veltrix_sdk/resources/tailscale.py
# -*- coding: utf-8 -*-

from .base import BaseResource


class TailscaleResource(BaseResource):
    """Handles Tailscale devices, keys and (admin) global configuration.

    Maps to ``/api/tailscale`` on the Community Edition server.
    """
    RESOURCE_PATH = "tailscale"

    def get_config(self, **kwargs):
        """Retrieves the tenant Tailscale config (GET /api/tailscale/config)."""
        return self._http_client.get(f"{self.RESOURCE_PATH}/config", **kwargs)

    def list_devices(self, **kwargs):
        """Lists Tailscale devices (GET /api/tailscale/devices)."""
        return self._http_client.get(f"{self.RESOURCE_PATH}/devices", **kwargs)

    def get_device(self, device_id, **kwargs):
        """Retrieves a single Tailscale device (GET /api/tailscale/devices/{id})."""
        return self._http_client.get(f"{self.RESOURCE_PATH}/devices/{device_id}", **kwargs)

    def generate_key(self, component_id, description, customer_id, reusable=None,
                     ephemeral=None, tags=None, **kwargs):
        """Generates a Tailscale auth key (POST /api/tailscale/keys)."""
        data = {
            "componentId": component_id,
            "description": description,
            "customerId": customer_id,
            "reusable": reusable,
            "ephemeral": ephemeral,
            "tags": tags,
        }
        data = {k: v for k, v in data.items() if v is not None}
        return self._http_client.post(f"{self.RESOURCE_PATH}/keys", data=data, **kwargs)

    def delete_device(self, device_id, **kwargs):
        """Deletes a Tailscale device (DELETE /api/tailscale/device/{id})."""
        return self._http_client.delete(f"{self.RESOURCE_PATH}/device/{device_id}", **kwargs)

    # --- Global config (admin only) ---

    def get_global_config(self, **kwargs):
        """Retrieves the global Tailscale config (GET /api/tailscale/global-config)."""
        return self._http_client.get(f"{self.RESOURCE_PATH}/global-config", **kwargs)

    def set_global_config(self, tailnet, api_key, api_url=None, **kwargs):
        """Sets the global Tailscale config (POST /api/tailscale/global-config)."""
        data = {"tailnet": tailnet, "apiKey": api_key, "apiUrl": api_url}
        data = {k: v for k, v in data.items() if v is not None}
        return self._http_client.post(f"{self.RESOURCE_PATH}/global-config", data=data, **kwargs)

    def delete_global_config(self, **kwargs):
        """Deletes the global Tailscale config (DELETE /api/tailscale/global-config)."""
        return self._http_client.delete(f"{self.RESOURCE_PATH}/global-config", **kwargs)

File: packages/sdk-js/src/resources/tailscale.ts
import { BaseResource } from './base';
import { AxiosRequestConfig } from 'axios';

// Define interfaces for Tailscale data structures
interface TailscaleDevice {
  id: string;
  name: string;
  hostname: string;
  user: string;
  addresses: string[];
  clientVersion: string;
  os: string;
  created: string; // ISO Date string
  lastSeen: string; // ISO Date string
  isOnline: boolean;
  tags?: string[];
}

interface TailscaleConfig {
    id: string;
    apiUrl: string;
    tailnet: string;
    apiKey: string; // Masked or omitted in responses usually
    enabled: boolean;
    createdAt: string; // ISO Date string
    updatedAt: string; // ISO Date string
}

interface GetConfigResponse {
    isConfigured: boolean;
    config?: TailscaleConfig | null;
}

interface GenerateKeyPayload {
    componentId: string; // uuid
    description: string;
    customerId: string; // uuid
    reusable?: boolean;
    ephemeral?: boolean;
    tags?: string[];
}

interface GenerateKeyResponse {
    key: string;
    expiresAt: string; // ISO Date string
}

export class TailscaleResource extends BaseResource {
  protected readonly RESOURCE_PATH = "tailscale";

  async getConfig(config?: AxiosRequestConfig): Promise<GetConfigResponse> {
    // Corresponds to GET /api/tailscale/config
    const path = `${this.RESOURCE_PATH}/config`;
    return this.httpClient.get<GetConfigResponse>(path, config);
  }

  async listDevices(config?: AxiosRequestConfig): Promise<TailscaleDevice[]> {
    // Corresponds to GET /api/tailscale/devices
    const path = `${this.RESOURCE_PATH}/devices`;
    return this.httpClient.get<TailscaleDevice[]>(path, config);
  }

  async getDevice(deviceId: string, config?: AxiosRequestConfig): Promise<TailscaleDevice> {
    // Corresponds to GET /api/tailscale/devices/{id}
    const path = `${this.RESOURCE_PATH}/devices/${deviceId}`;
    return this.httpClient.get<TailscaleDevice>(path, config);
  }

  async generateKey(payload: GenerateKeyPayload, config?: AxiosRequestConfig): Promise<GenerateKeyResponse> {
    // Corresponds to POST /api/tailscale/keys
    const path = `${this.RESOURCE_PATH}/keys`;
     const cleanedPayload = Object.entries(payload)
        .filter(([_, value]) => value !== undefined) // Keep nulls if intended
        .reduce((acc, [key, value]) => ({ ...acc, [key]: value }), {});
    return this.httpClient.post<GenerateKeyResponse>(path, cleanedPayload, config);
  }

  async deleteDevice(deviceId: string, config?: AxiosRequestConfig): Promise<null> {
    // Corresponds to DELETE /api/tailscale/device/{id}
    // Note singular 'device' in path
    const path = `${this.RESOURCE_PATH}/device/${deviceId}`;
    // API Spec indicates 204 No Content
    return this.httpClient.delete<null>(path, config);
  }

  // --- Global Config Methods (Admin Only) ---

  async getGlobalConfig(config?: AxiosRequestConfig): Promise<TailscaleConfig> {
    // Corresponds to GET /api/tailscale/global-config
    const path = `${this.RESOURCE_PATH}/global-config`;
    return this.httpClient.get<TailscaleConfig>(path, config);
  }

  async setGlobalConfig(payload: { tailnet: string; apiKey: string; apiUrl?: string }, config?: AxiosRequestConfig): Promise<TailscaleConfig> {
    // Corresponds to POST /api/tailscale/global-config
    const path = `${this.RESOURCE_PATH}/global-config`;
     const cleanedPayload = Object.entries(payload)
        .filter(([_, value]) => value !== undefined) // Keep nulls if intended
        .reduce((acc, [key, value]) => ({ ...acc, [key]: value }), {});
    return this.httpClient.post<TailscaleConfig>(path, cleanedPayload, config);
  }

  async deleteGlobalConfig(config?: AxiosRequestConfig): Promise<null> {
    // Corresponds to DELETE /api/tailscale/global-config
    // API Spec indicates 204 No Content
    const path = `${this.RESOURCE_PATH}/global-config`;
    return this.httpClient.delete<null>(path, config);
  }
}
